File: csrc/csr_gws.cpp
#include "./cuda/header_cuda.h"
#include <ATen/Parallel.h>
#include <ATen/core/Tensor.h>
#include <torch/extension.h>
#include <torch/library.h>

using torch::autograd::AutogradContext;
using torch::autograd::Variable;
using torch::autograd::variable_list;

TORCH_LIBRARY_FRAGMENT(geot, m) {
  m.def("csr_gws_impl(Tensor indptr, Tensor indices, Tensor "
        "weight, Tensor src) -> Tensor");
  m.def("sddmm_csr_impl(Tensor rowptr, Tensor colind, Tensor mat_1, "
        "Tensor mat_2) -> Tensor");
  m.def("csr2csc_impl(Tensor csr_ptr, Tensor csr_ind, Tensor csr_val, "
        "Tensor csc_ptr, Tensor csc_ind, Tensor csc_val) -> void");
}

// this kernel take a set of sorted csr tensors and scatter the src tensor
// mainly apply for GNN operation
// so the tensor is 2D, indptr is a 1D tensor of size nrow + 1
// indices is a 1D tensor of size nnz, weight is a 1D tensor of size nnz
at::Tensor csr_gws_cuda_fwd_impl(at::Tensor indptr, at::Tensor indices,
                                 at::Tensor weight, at::Tensor src) {
  // convert dtype to int32
  at::Tensor indptr_int = indptr.to(torch::kInt32);
  at::Tensor indices_int = indices.to(torch::kInt32);
  auto max_index = indptr.size(0) - 1;
  auto output_shape = src.sizes().vec();
  output_shape[0] = max_index;
  auto output = torch::zeros(output_shape, src.options());
  csr_gws_cuda(indptr_int, indices_int, weight, src, output);
  return output;
}

at::Tensor sddmm_csr_cuda_fwd_impl(at::Tensor rowptr, at::Tensor colind,
                                   at::Tensor mat_1, at::Tensor mat_2) {
  // rowptr = rowptr.to(torch::kInt32);
  // colind = colind.to(torch::kInt32);
  auto output_shape = {colind.size(0)};
  auto output = torch::zeros(output_shape, mat_1.options());
  sddmm_csr_cuda(rowptr, colind, mat_1, mat_2, output);
  return output;
}

void csr2csc_cuda_fwd_impl(at::Tensor csrptr, at::Tensor csrind,
                           at::Tensor csrval, at::Tensor cscptr,
                           at::Tensor cscind, at::Tensor cscval) {
  // convert dtype to int32
  at::Tensor csrptr_int = csrptr.to(torch::kInt32);
  at::Tensor csrind_int = csrind.to(torch::kInt32);
  csr2csc_cuda(csrptr_int, csrind_int, csrval, cscptr, cscind, cscval);
}

TORCH_LIBRARY_IMPL(geot, CUDA, m) {
  m.impl("csr_gws_impl", &csr_gws_cuda_fwd_impl);
  m.impl("sddmm_csr_impl", &sddmm_csr_cuda_fwd_impl);
  m.impl("csr2csc_impl", &csr2csc_cuda_fwd_impl);
}

File: geot/csr_gws.py
import torch

def csr_gws_impl(csrptr: torch.Tensor, csrind: torch.Tensor, weight: torch.Tensor, 
                               src: torch.Tensor) -> torch.Tensor:
    return torch.ops.geot.csr_gws_impl(csrptr, csrind, weight, src)

# def sddmm_csr_impl(csrptr: torch.Tensor, csrind: torch.Tensor, 
#                mat_1: torch.Tensor, mat_2: torch.Tensor) -> torch.Tensor:
#     csrptr = csrptr.to(torch.int32)
#     csrind = csrind.to(torch.int32)
#     return torch.ops.geot.sddmm_csr_impl(csrptr, csrind, mat_1, mat_2)

# def csr2csc_impl(csrptr: torch.Tensor, csrind: torch.Tensor, csrval: torch.Tensor) -> torch.Tensor:
#     m = csrptr.shape[0] - 1
#     n = csrind.max() + 1
#     nnz = csrind.shape[0]
    
#     cscptr = torch.zeros(n+1, dtype=torch.int32, device=csrptr.device)
#     cscind = torch.zeros(nnz, dtype=torch.int32, device=csrind.device)
#     cscval = torch.zeros(nnz, dtype=csrval.dtype, device=csrval.device)
#     torch.ops.geot.csr2csc_impl(csrptr, csrind, csrval, cscptr, cscind, cscval)
#     return cscptr, cscind, cscval


@torch.library.custom_op("geot::csr_gws", mutates_args=())
def csr_gws(csrptr: torch.Tensor, csrind: torch.Tensor, weight: torch.Tensor,
                          src: torch.Tensor) -> torch.Tensor:
    return csr_gws_impl(csrptr, csrind, weight, src)

@torch.library.register_fake("geot::csr_gws")
def _(csrptr: torch.Tensor, csrind: torch.Tensor, weight: torch.Tensor, 
      src: torch.Tensor,) -> torch.Tensor:
    ctx = torch.library.get_ctx()
    dst_node = ctx.new_dynamic_size()
    shape = [dst_node, src.shape[1]]
    out = torch.zeros(shape, dtype=src.dtype, device=src.device)
    return out

# def setup_context(ctx, inputs, output) -> torch.Tensor:
#     csrptr, csrind, weight, src = inputs
#     ctx.save_for_backward(csrptr, csrind, weight, src)
    
# def backward(ctx, grad):
#     csrptr, csrind, weight, src = ctx.saved_tensors
#     grad = grad.contiguous()
    
#     csc_ptr, csc_ind, csc_val = csr2csc_impl(csrptr, csrind, weight)
    
#     src_grad = csr_gws_impl(csc_ptr, csc_ind, csc_val, grad)
#     weight_grad = sddmm_csr_impl(csc_ptr, csc_ind, grad, src)

#     return (None, None, weight_grad, src_grad)

# torch.library.register_autograd("geot::csr_gws", backward, setup_context=setup_context)